### PR TITLE
Move drs resolution during copy into background

### PR DESCRIPTION
**Pull request by xbrianh**
Previously DRS resolution was performed serially in the main
thread/process. This moves DRS resolution into the background,
improving batch copy performance.

### CODE REVIEW

File: terra_notebook_utils/drs.py [hunk 1]
```diff
@@ -200,50 +201,62 @@ def _resolve_local_target(filepath: str, info: DRSInfo) -> str:
...
-                  workspace_name: Optional[str]=WORKSPACE_NAME,
-                  workspace_namespace: Optional[str]=WORKSPACE_GOOGLE_PROJECT):
-    """Copy a DRS object to the local filesystem."""
+def _do_copy_drs(drs_uri: str,
+                 dst: str,
+                 multipart_threshold: int,
+                 workspace_name: Optional[str]=WORKSPACE_NAME,
+                 workspace_namespace: Optional[str]=WORKSPACE_GOOGLE_PROJECT):
+    dst_blob: Union[GSBlob, URLBlob, LocalBlob]
+    if dst.startswith("gs://"):
```
Comment (DailyDreaming) [inline terra_notebook_utils/drs.py:210]: DRY suggestion:

```
if dst.startswith("gs://"):
    src_info = get_drs_info(drs_uri)
    src_blob = get_drs_blob(src_info, workspace_namespace)
    bucket_name, key = _resolve_bucket_target(dst, src_info)
    dst_blob = GSBlob(bucket_name, key)
    copy_client._do_copy(src_blob, dst_blob, multipart_threshold)
else:
    info = get_drs_info(drs_uri)
    src_blob = get_drs_blob(info, workspace_namespace)
    dst_blob = copy_client.blob_for_url(_resolve_local_target(dst, info))
    copy_client._do_copy(src_blob, dst_blob, multipart_threshold)
```

->

```
src_info = get_drs_info(drs_uri)
src_blob = get_drs_blob(src_info, workspace_namespace)
if dst.startswith("gs://"):
    bucket_name, key = _resolve_bucket_target(dst, src_info)
    dst_blob = GSBlob(bucket_name, key)
else:
    dst_blob = copy_client.blob_for_url(_resolve_local_target(dst, src_info))
copy_client._do_copy(src_blob, dst_blob, multipart_threshold)
```

File: terra_notebook_utils/drs.py [hunk 2]
```diff
@@ -200,50 +201,62 @@ def _resolve_local_target(filepath: str, info: DRSInfo) -> str:
...
+    with DRSCopyClient(raise_on_error=True, progress_indicator="bar") as cc:
+        cc.workspace = workspace_name
+        cc.workspace_namespace = workspace_namespace
+        cc.copy(drs_uri, dst or ".")
 
-def copy_to_bucket(drs_url: str,
-                   dst_key: str,
+def copy_to_bucket(drs_uri: str,
+                   dst_key: str="",
                    dst_bucket_name: str=None,
```
Comment (DailyDreaming) [inline terra_notebook_utils/drs.py:249]: Not part of the PR but `Optional[str]`.

**Review by DailyDreaming — APPROVED** [review]
LGTM.

File: terra_notebook_utils/drs.py (hunk 1 above)
Comment (xbrianh) [reply terra_notebook_utils/drs.py:210]: thanks! will fix.

File: terra_notebook_utils/drs.py (hunk 2 above)
Comment (xbrianh) [reply terra_notebook_utils/drs.py:249]: An empty string seems fine here.

Comment (DailyDreaming) [reply terra_notebook_utils/drs.py:247]: I agree, I meant for the one under it that had a `None` already.
